core/iwasm/common/wasm_runtime_common.c

```diff
@@ -2902,6 +2905,10 @@ wasm_runtime_destroy_wasi(WASMModuleInstanceCommon *module_inst)
 
         wasm_runtime_free(wasi_ctx);
     }
+
+#if WASM_ENABLE_WASI_NN != 0
+    wasi_nn_destroy();
```

Comment:
please move to unregister native api in wasi_native.c，ref to calling of lib_pthread_destroy

Reply:
@wenyongh, I added `wasi_nn_destroy` because I wanted to free the mem allocated by the WASM app when unload. Here it seems that it will be done when tearing down the runtime. Is that correct?

Reply:
@tonibofarull Do you mean that the wasm app calls function `wasi_nn_load` to load and allocate resources, while it won't call `wasi_nn_unload` like function to unload and free the resources?
If yes, there may be two methods to resolve it: (1) wasi_nn_load firstly checks whether there are resources allocated by last wasm app and free them if found (like calling wasi_nn_destroy), and then allocates memory for current wasm app. And when runtime exits, it calls wasi_nn_destroy inside wasm_native_destroy to free the resources if found. (2) store the allocated resources, e.g. add them into a list, and when runtime exits, in wasi_nn_destroy, traverse the list to free the resources.
Seems the first one is better, but there should be no two wasm apps running simultaneously in two threads.

Reply:
We can apply method 1. But what happens if we load 2 WASM apps that use wasi-nn at the same time? I was planning to allocate resources separately and deallocate once the apps are unloaded.

Reply:
NOTE: Right now there is an implementation restriction of only running wasi-nn in 1 WASM app, but this limitations is because of the current implementation and no API specific.

Reply:
Another way is to store the wasi-nn resources of each wasm app into its wasm module instance, for example, add a field pointing to a context of wasi-nn (e.g. `wasi_nn_ctx` similar to `wasi_ctx` for libc-wasi) in `WASM/AOTModuleInstanceExtra`, create the context when instantiation (in `wasm_instantiate` and `aot_instantiate`) and manipulate it in `wasi_nn_load`, and destroy the context when deinstantiation (in `wasm_deinstantiate` and `aot_deinstantiate`).

Maybe we can document the limitation in this version and merge this PR firstly? And optimize it later?

Reply:
Fixed here: https://github.com/midokura/wasm-micro-runtime/commit/fa6c5a4fe594e1ca8826c87cefab1e8061a2a0af